### PR TITLE
cmark extensions

### LINKED ISSUE #22: No rule to make target libcmark_static
On Windows 10, trying to install commonmarker with `gem install commonmarker`, I get this:

```
C:/Ruby22-x64/bin/ruby.exe -r ./siteconf20160430-2356-a8w5ay.rb extconf.rb
checking for cmake... yes
-- Building for: Visual Studio 14 2015
-- The C compiler identification is MSVC 19.0.23918.0
-- The CXX compiler identification is MSVC 19.0.23918.0
-- Check for working C compiler using: Visual Studio 14 2015
-- Check for working C compiler using: Visual Studio 14 2015 -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler using: Visual Studio 14 2015
-- Check for working CXX compiler using: Visual Studio 14 2015 -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test HAVE_FLAG_ADDRESS_SANITIZER
-- Performing Test HAVE_FLAG_ADDRESS_SANITIZER - Failed
-- Performing Test HAVE_FLAG_SANITIZE_ADDRESS
-- Performing Test HAVE_FLAG_SANITIZE_ADDRESS - Failed
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR - Failed
-- Performing Test COMPILER_HAS_DEPRECATED
-- Performing Test COMPILER_HAS_DEPRECATED - Success
-- Looking for stdbool.h
-- Looking for stdbool.h - found
-- Performing Test HAVE___BUILTIN_EXPECT
-- Performing Test HAVE___BUILTIN_EXPECT - Failed
-- Performing Test HAVE___ATTRIBUTE__
-- Performing Test HAVE___ATTRIBUTE__ - Failed
-- Found PythonInterp: C:/Users/Louis/AppData/Local/Programs/Python/Python35-32/python.exe (found suitable version "3.5.1", minimum required is "3") 
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Ruby22-x64/lib/ruby/gems/2.2.0/gems/commonmarker-0.8.0/ext/commonmarker/cmark/build
make: *** No rule to make target `libcmark_static'.  Stop.
checking for cmark_parse_document() in -lcmark... no
libcmark is missing.
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--without-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=C:/Ruby22-x64/bin/$(RUBY_BASE_NAME)
	--with-cmark-dir
	--without-cmark-dir
	--with-cmark-include
	--without-cmark-include=${cmark-dir}/include
	--with-cmark-lib
	--without-cmark-lib=${cmark-dir}/lib
	--with-cmarklib
	--without-cmarklib

extconf failed, exit code 1
```

The mkmf.log file says the following:
```
find_executable: checking for cmake... -------------------- yes

--------------------

find_library: checking for cmark_parse_document() in -lcmark... -------------------- no

"x86_64-w64-mingw32-gcc -o conftest.exe -IC:/Ruby22-x64/include/ruby-2.2.0/x64-mingw32 -IC:/Ruby22-x64/include/ruby-2.2.0/ruby/backward -IC:/Ruby22-x64/include/ruby-2.2.0 -I. -IC:/Ruby22-x64/include -DFD_SETSIZE=2048 -D_WIN32_WINNT=0x0501 -D__MINGW_USE_VC2005_COMPAT -D_FILE_OFFSET_BITS=64   -O3 -fno-omit-frame-pointer -fno-fast-math -g -Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wno-long-long -Wno-missing-field-initializers -Wunused-variable -Wpointer-arith -Wwrite-strings -Wdeclaration-after-statement -Wimplicit-function-declaration -Wdeprecated-declarations -Wno-packed-bitfield-compat conftest.c  -LC:/Ruby22-x64/lib -LC:/Ruby22-x64/lib/ruby/gems/2.2.0/gems/commonmarker-0.8.0/ext/commonmarker/cmark/build/src -L. -LC:/Ruby22-x64/lib -L.      -lx64-msvcrt-ruby220  -lshell32 -lws2_32 -liphlpapi -limagehlp -lshlwapi  "
checked program was:
/* begin */
1: #include "ruby.h"
2: 
3: #include <winsock2.h>
4: #include <windows.h>
5: int main(int argc, char **argv)
6: {
7:   return 0;
8: }
/* end */

"x86_64-w64-mingw32-gcc -o conftest.exe -IC:/Ruby22-x64/include/ruby-2.2.0/x64-mingw32 -IC:/Ruby22-x64/include/ruby-2.2.0/ruby/backward -IC:/Ruby22-x64/include/ruby-2.2.0 -I. -IC:/Ruby22-x64/include -DFD_SETSIZE=2048 -D_WIN32_WINNT=0x0501 -D__MINGW_USE_VC2005_COMPAT -D_FILE_OFFSET_BITS=64   -O3 -fno-omit-frame-pointer -fno-fast-math -g -Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wno-long-long -Wno-missing-field-initializers -Wunused-variable -Wpointer-arith -Wwrite-strings -Wdeclaration-after-statement -Wimplicit-function-declaration -Wdeprecated-declarations -Wno-packed-bitfield-compat conftest.c  -LC:/Ruby22-x64/lib -LC:/Ruby22-x64/lib/ruby/gems/2.2.0/gems/commonmarker-0.8.0/ext/commonmarker/cmark/build/src -L. -LC:/Ruby22-x64/lib -L.      -lx64-msvcrt-ruby220 -lcmark  -lshell32 -lws2_32 -liphlpapi -limagehlp -lshlwapi  "
conftest.c: In function 't':
conftest.c:16:57: error: 'cmark_parse_document' undeclared (first use in this function)
conftest.c:16:57: note: each undeclared identifier is reported only once for each function it appears in
conftest.c:16:32: warning: variable 'p' set but not used [-Wunused-but-set-variable]
checked program was:
/* begin */
 1: #include "ruby.h"
 2: 
 3: #include <winsock2.h>
 4: #include <windows.h>
 5: 
 6: /*top*/
 7: extern int t(void);
 8: int main(int argc, char **argv)
 9: {
10:   if (argc > 1000000) {
11:     printf("%p", &t);
12:   }
13: 
14:   return 0;
15: }
16: int t(void) { void ((*volatile p)()); p = (void ((*)()))cmark_parse_document; return 0; }
/* end */

"x86_64-w64-mingw32-gcc -o conftest.exe -IC:/Ruby22-x64/include/ruby-2.2.0/x64-mingw32 -IC:/Ruby22-x64/include/ruby-2.2.0/ruby/backward -IC:/Ruby22-x64/include/ruby-2.2.0 -I. -IC:/Ruby22-x64/include -DFD_SETSIZE=2048 -D_WIN32_WINNT=0x0501 -D__MINGW_USE_VC2005_COMPAT -D_FILE_OFFSET_BITS=64   -O3 -fno-omit-frame-pointer -fno-fast-math -g -Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wno-long-long -Wno-missing-field-initializers -Wunused-variable -Wpointer-arith -Wwrite-strings -Wdeclaration-after-statement -Wimplicit-function-declaration -Wdeprecated-declarations -Wno-packed-bitfield-compat conftest.c  -LC:/Ruby22-x64/lib -LC:/Ruby22-x64/lib/ruby/gems/2.2.0/gems/commonmarker-0.8.0/ext/commonmarker/cmark/build/src -L. -LC:/Ruby22-x64/lib -L.      -lx64-msvcrt-ruby220 -lcmark  -lshell32 -lws2_32 -liphlpapi -limagehlp -lshlwapi  "
conftest.c: In function 't':
conftest.c:16:1: warning: implicit declaration of function 'cmark_parse_document' [-Wimplicit-function-declaration]
c:/rubydevkit/mingw/bin/../lib/gcc/x86_64-w64-mingw32/4.7.2/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lcmark
collect2.exe: error: ld returned 1 exit status
checked program was:
/* begin */
 1: #include "ruby.h"
 2: 
 3: #include <winsock2.h>
 4: #include <windows.h>
 5: 
 6: /*top*/
 7: extern int t(void);
 8: int main(int argc, char **argv)
 9: {
10:   if (argc > 1000000) {
11:     printf("%p", &t);
12:   }
13: 
14:   return 0;
15: }
16: int t(void) { cmark_parse_document(); return 0; }
/* end */

--------------------
```

### PR DESCRIPTION
**Pull request by kivikakk**
This is not yet ready for merge, but I figure opening a PR is an okay idea at this stage.

In short:

* it uses my `extensions` branch of cmark: https://github.com/kivikakk/cmark/tree/extensions
  * this adds extensibility based on https://github.com/jgm/cmark/pull/123, but in such a way that the extensions do not require changes to the core code
  * it also doesn't do any searching and autoloading of shared objects to find extensions (!)
* we register all the core extensions included in this branch
* the commonmarker user can then elect which extensions to enable at parse time
* added rendering back to commonmark!

Fixes #22.

### CODE REVIEW

**Comment by gjtorikian** [comment]
Not sure if you're aware but it's totally fine for the Windows build to fail: https://github.com/gjtorikian/commonmarker/issues/22

I should get on that....one day. 😞 

Other than that, this is looking great. The two things that stand out are:

* obviously, all the `XXX should raise` comments
* I prefer to use symbols for immutable strings, so if `extensions` could be an array of symbols, instead, that'd be great. 

I'm not nervous about switching the submodule to your fork. I _am_ nervous about the core CMark stuff getting improvements, that might not trickle down to your fork. For example, I routinely run [`update_submodule`](https://github.com/gjtorikian/commonmarker/blob/fc173dfbaf9b8c0d9325d4c6fc58a000a2aa564c/script/update_submodules) on a new CMark release and just bump the gem: https://github.com/gjtorikian/commonmarker/pull/21

Could/Would you be willing to also incorporate those upstream C changes into your fork, so that I can continue to release new gems with those updates? If not, I'm totally open to any other suggestions or improvements on the workflow here. For example, could any of the C extension code be written directly to this repo, rather than maintained in a separate submodule?

**Comment by kivikakk** [comment]
Haha, whoops! If I ever get a Windows dev environment going I might look into it.

I could've sworn I had a good reason to use symbols instead of strings — maybe it was because one of the extension names had a hyphen in it. At any rate, I'll change it to use symbols there; will look much better!

I'm super happy to track the upstream cmark and integrate — hopefully it'll only be a matter of time before this code does get merged upstream, so it'll be helpful for me to do this anyway, to make sure it always stays merge-ready. :)

Unfortunately the nature of the changes (adapting cmark to _be_ extensible) means we can't easily separate it yet. I'm hoping the core changes will land, and then we'll be able to keep the extensions proper in this repo if we wanted/needed.

**Comment by kivikakk** [comment]
Symbols are in, `XXX` are out, upstream is merged. :)

**Comment by kivikakk** [comment]
btw, Windows build failures currently look a little different to #22; might be http://help.appveyor.com/discussions/problems/4294-all-builds-say-specify-a-project-or-solution-file-the-directory-does-not-contain-a-project-or-solution-file ?

**Comment by kivikakk** [comment]
Windows is all fixed. Let me know if there's anything else you'd like addressed!

**Comment by gjtorikian** [comment]
😹

File: appveyor.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+build: off
+
+before_test:
+  - ruby -v
+  - gem -v
+  - bundle -v
+
+test_script:
+  - bundle exec rake
```
Comment (gjtorikian) [inline appveyor.yml]: OCD ALERT: can this end in a newline please? 🙏

File: lib/commonmarker.rb [hunk 2]
```diff
@@ -15,26 +15,28 @@ module CommonMarker
   #
   # text - A {String} of text
   # option - Either a {Symbol} or {Array of Symbol}s indicating the render options
+  # extensions - An {Array of String}s indicating the extensions to use
```
Comment (gjtorikian) [inline lib/commonmarker.rb]: Should be `{Array of Symbol}s`.

File: lib/commonmarker.rb [hunk 3]
```diff
@@ -15,26 +15,28 @@ module CommonMarker
...
-    html = Node.markdown_to_html(text, opts)
+    html = Node.markdown_to_html(text, opts, extensions)
     html.force_encoding('UTF-8')
   end
 
   # Public: Parses a Markdown string into a `document` node.
   #
   # string - {String} to be parsed
   # option - A {Symbol} or {Array of Symbol}s indicating the parse options
+  # extensions - An {Array of String}s indicating the extensions to use
```
Comment (gjtorikian) [inline lib/commonmarker.rb]: Should be `{Array of Symbol}s`.

File: lib/commonmarker/node.rb [hunk 4]
```diff
@@ -17,11 +17,22 @@ def walk(&block)
     # Public: Convert the node to an HTML string.
     #
     # options - A {Symbol} or {Array of Symbol}s indicating the render options
+    # extensions - An {Array of String}s indicating the extensions to use
```
Comment (gjtorikian) [inline lib/commonmarker/node.rb]: Should be `{Array of Symbol}s`.

File: test/test_extensions.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    none_in_use = CommonMarker.render_html(@markdown, :default, %i[table strikethrough])
+    refute none_in_use.include?("| a")
+    refute none_in_use.include?("~~hi~~")
+  end
+
+  def test_bad_extension_specifications
+    assert_raises(TypeError) { CommonMarker.render_html(@markdown, :default, "nope") }
+    assert_raises(TypeError) { CommonMarker.render_html(@markdown, :default, ["table"]) }
+    assert_raises(ArgumentError) { CommonMarker.render_html(@markdown, :default, %i[table bad]) }
+  end
```
Comment (gjtorikian) [inline test/test_extensions.rb]: Love the error cases--thank you!

File: test/test_extensions.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    refute none_in_use.include?("| a")
+    assert none_in_use.include?("~~hi~~")
+
+    none_in_use = CommonMarker.render_html(@markdown, :default, %i[strikethrough])
+    assert none_in_use.include?("| a")
+    refute none_in_use.include?("~~hi~~")
+
+    none_in_use = CommonMarker.render_html(@markdown, :default, %i[table strikethrough])
+    refute none_in_use.include?("| a")
+    refute none_in_use.include?("~~hi~~")
```
Comment (gjtorikian) [inline test/test_extensions.rb]: Could we get some assertions on the expected HTML, for example, `<strike>hi</strike>` or `<td>a</td>`? (For these, and the tests above.)

Also, could you demonstrate what happens with something like, asterisks in a table cell? `github-markdown` does support this at the moment.

**Comment by gjtorikian** [comment]
This is looking goooood! I left a few more comments. Thanks for your patience.

File: test/test_extensions.rb (hunk 6 above)
Comment (kivikakk) [inline test/test_extensions.rb]: Can do! I had left these out as cmark itself tests the cases, but I think it makes sense to assert they're working properly here too.

**Comment by kivikakk** [comment]
Fixed up. And not at all, thanks for your attention! :)

**Comment by gjtorikian** [comment]
Shipping as 0.11.0!

**Pull request merged**